### PR TITLE
Shifted boundary method (Dirichlet)

### PR DESCRIPTION
**Pull request by kmittal2**
This PR introduces a miniapp containing a high-order implementation of the [second-generation](https://www.sciencedirect.com/science/article/pii/S0045782520305260) Shifted Boundary Method for the Laplace equation with Dirichlet BC.

Example: Dirichlet BC are imposed on the zero level set of a given function, on a uniform mesh. The algorithm uses a distance function to the true boundary to force Dirichlet BC on the (non-aligned) mesh faces, therefore "shifting" the location where BC are imposed.

Screen Shot 2021-03-31 at 10 41 32 PM

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2043](https://github.com/mfem/mfem/pull/2043) | @kmittal2 | @tzanio | @vladotomov + @bslazarov + @dylan-copeland | 04/01/21 | 05/25/21 | 06/04/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/linearform.cpp [hunk 1]
```diff
@@ -92,14 +114,31 @@ void LinearForm::Assemble()
...
+                        dlfi_elem_attr_marker[k]->Size(),
+                        "invalid element marker for domain linear form "
+                        "integrator #" << k << ", counting from zero");
+         }
+      }
+
       for (i = 0; i < fes -> GetNE(); i++)
       {
+         int elem_attr = fes->GetMesh()->GetAttribute(i);
          fes -> GetElementVDofs (i, vdofs);
```
Comment (vladotomov) [inline fem/linearform.cpp:131]: Same as in `bilinearform`:
It would be good to skip all computations (`GetElementDofs`, `GetElementTransformation` etc) if the element's attribute is not active for any integrator. E.g., see the logic with `bdr_attr_marker` below.

**Review by vladotomov — CHANGES_REQUESTED** [review]

File: miniapps/shifted/diffusion.cpp [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+//      -----------------------------------------------------------------
+//      Shifted Boundary Miniapp: Finite element immersed boundary solver
+//      -----------------------------------------------------------------
+//
+// TODO - problem and approach descriptions, references.
```
Comment (vladotomov) [inline miniapps/shifted/diffusion.cpp:16]: Please add the above.

File: miniapps/shifted/diffusion.cpp [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+//   Problem 1: Circular hole of radius 0.2 at the center of the domain.
+//              Solves -nabla^u = 1 with homogeneous boundary conditions.
+//   mpirun -np 4 diffusion -m ../../data/inline-quad.mesh -rs 3 -o 1 -vis -lst 1
+//   mpirun -np 4 diffusion -m ../../data/inline-hex.mesh -rs 2 -o 2 -vis -lst 1 -ho 1 -alpha 10
+//
+//   Problem 2: Circular hole of radius 0.2 at the center of the domain.
+//              Solves -nabla^u = f with inhomogeneous boundary conditions, and
+//              f is setup such that u = x^p + y^p, where p = 2 by default.
+//              This is a 2D convergence test.
+//   mpirun -np 4 diffusion -rs 2 -o 2 -vis -lst 2
```
Comment (vladotomov) [inline miniapps/shifted/diffusion.cpp:30]: Could you add more details how to run the convergence tests. I don't think simply changing `-rs` works.

**Review by vladotomov — CHANGES_REQUESTED** [review]

**Comment by kmittal2** [comment]
Companion web PR: https://github.com/mfem/web/pull/103, https://github.com/GLVis/web/pull/5.

**Review by vladotomov — APPROVED** [review]

**Comment by tzanio** [comment]
@bslazarov and @rw-anderson -- Can you please review? This is now overdue.

**Review by bslazarov — APPROVED** [review]
Excellent! The code is tested for all of the provided test cases - works great and is ready to be merged.

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: miniapps/shifted/sbm-aux.hpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
+﻿//                       MFEM
+//
+// Compile with: make ex1p
```
Comment (dylan-copeland) [inline miniapps/shifted/sbm-aux.hpp:3]: Remove these comments.

File: miniapps/shifted/sbm-aux.hpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
+﻿//                       MFEM
```
Comment (dylan-copeland) [inline miniapps/shifted/sbm-aux.hpp:1]: The copyright stuff should be copied here.

File: miniapps/shifted/sbm-aux.hpp [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+// mpirun -np 1 ex1p -m ../../data/inline-quad.mesh  -rs 0 -vis -o 2
+
+#include "../../mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+/// Analytic distance to the 0 level set. +ve value if the point is inside the
```
Comment (dylan-copeland) [inline miniapps/shifted/sbm-aux.hpp:15]: What is `ve`?

File: miniapps/shifted/marking.hpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+   /// If @a include_cut_cell = true, the surrogate boundary is the faces
+   /// between elements outside the true domain and the elements cut by the true
+   /// boundary.
+   void ListShiftedFaceDofs(const Array<int> &elem_marker,
+                            Array<int> &sface_dof_list) const;
+
+   /// List the dofs that will be inactive for the computation on the surrogate
+   /// domain. This include dofs for the elements located outside the
+   /// true domain (and optionally, for the elements cut by the true boundary,
+   /// if @a include_cut_cell = true) minus the dofs that are located on the
```
Comment (dylan-copeland) [inline miniapps/shifted/marking.hpp:63]: Should it be `include_cut_cell == false`?

File: miniapps/shifted/marking.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+   /// between elements cut by the true boundary and the elements that are
+   /// located inside the true domain.
+   /// If @a include_cut_cell = true, the surrogate boundary is the faces
+   /// between elements outside the true domain and the elements cut by the true
+   /// boundary.
+   void ListShiftedFaceDofs(const Array<int> &elem_marker,
+                            Array<int> &sface_dof_list) const;
+
+   /// List the dofs that will be inactive for the computation on the surrogate
+   /// domain. This include dofs for the elements located outside the
```
Comment (dylan-copeland) [inline miniapps/shifted/marking.hpp:61]: include -> includes

File: miniapps/shifted/sbm_solver.hpp [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+               const IntegrationPoint &ip,
+               const Vector &D);
+};
+
+class SBM2DirichletIntegrator : public BilinearFormIntegrator
+{
+protected:
+   double alpha;
+   VectorCoefficient *vD; // Distance function coefficient
+   Array<int> *elem_marker; //marker indicating wether element is inside,
```
Comment (dylan-copeland) [inline miniapps/shifted/sbm_solver.hpp:50]: wether -> whether (occurs again below)

File: miniapps/shifted/sbm_solver.cpp [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+      else
+      {
+         el2.CalcShape(eip2, shape);
+         el2.CalcDShape(eip2, dshape);
+         hinvdx = nor*nor/Tr.Elem2->Weight();
+         w = ip.weight * uD->Eval(Tr, ip, D) / Tr.Elem2->Weight();
+         CalcAdjugate(Tr.Elem2->Jacobian(), adjJ);
+      }
+
+
```
Comment (dylan-copeland) [inline miniapps/shifted/sbm_solver.cpp:578]: extra line

File: fem/lininteg.cpp [hunk 11]
```diff
@@ -23,6 +23,14 @@ void LinearFormIntegrator::AssembleRHSElementVect(
...
 
 
+void LinearFormIntegrator::AssembleRHSElementVect(
+   const FiniteElement &el1, const FiniteElement &el2,
+   FaceElementTransformations &Tr, Vector &elvect)
+{
+   mfem_error("LinearFormIntegrator::AssembleRHSElementVect(...)");
+}
+
+
```
Comment (dylan-copeland) [inline fem/lininteg.cpp:33]: extra line (also before this function)

File: fem/linearform.cpp [hunk 12]
```diff
@@ -194,6 +232,32 @@ void LinearForm::Assemble()
          }
       }
    }
+
+
```
Comment (dylan-copeland) [inline fem/linearform.cpp:236]: extra line

File: miniapps/shifted/sbm_solver.cpp [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+   dim = el1.GetDim();
+   ndof1 = el1.GetDof();
+   ndof2 = el2.GetDof();
+   ndoftotal = Trans.ElementType == ElementTransformation::BDR_FACE ?
+               ndof1 : ndof1 + ndof2;
+
+   elmat.SetSize(ndoftotal);
+   elmat = 0.0;
+
+   bool elem1f = true;
```
Comment (dylan-copeland) [inline miniapps/shifted/sbm_solver.cpp:50]: This function is hard to follow and could use more documentation. If someone wants to use this miniapp as the basis for doing SBM for a more complicated application, wouldn't they need to understand the details here rather than just using this function? If so, it should be well documented. It's not clear to me whether this is something internal that users won't need to understand.

**Comment by dylan-copeland** [comment]
Not a major issue, but there is an inconsistency in file names in `miniapps/shifted`. Most files have underscores, but `sbm-aux.hpp` has a hyphen. I don't think it needs to be consistent, but if you want to change it, now would be the best time.

File: miniapps/shifted/diffusion.cpp [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+      int  visport   = 19916, s = 350;
+      socketstream sol_sock;
+      common::VisualizeField(sol_sock, vishost, visport, face_dofs,
+                             "Shifted Face Dofs", 0, s, s, s, "Rjmp");
+   }
+
+   // Make a list of inactive tdofs that will be eliminated from the system.
+   // The inactive tdofs are the dofs for the elements located outside the
+   // true domain (and optionally, for the elements cut by the true boundary,
+   // if include_cut_cell = true) minus the dofs that are located on the
```
Comment (dylan-copeland) [inline miniapps/shifted/diffusion.cpp:225]: `true` -> `false`?

File: miniapps/shifted/marking.cpp [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+   {
+      FaceElementTransformations *tr = pmesh.GetSharedFaceTransformations(i);
+      if (tr != NULL)
+      {
+         int ne1 = tr->Elem1No;
+         int te1 = elem_marker[ne1];
+         int te2 = elem_marker[i+pmesh.GetNE()];
+         const int faceno = pmesh.GetSharedFace(i);
+         // Add if the element on this proc is completely inside the domain
+         // and the the element on other proc is not
```
Comment (dylan-copeland) [inline miniapps/shifted/marking.cpp:159]: typo

File: miniapps/shifted/marking.cpp [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+      }
+   }
+
+   // Combine the lists to mark essential dofs.
+   for (int i = 0; i < ess_vdofs.Size(); i++)
+   {
+      if (ess_vdofs_bdr[i] == -1) { ess_vdofs[i] = -1; }
+   }
+
+   // Unmark dofs that are on SBM faces (but not on dirichlet boundaries)
```
Comment (dylan-copeland) [inline miniapps/shifted/marking.cpp:264]: Dirichlet

**Review by dylan-copeland — APPROVED** [review]

File: miniapps/shifted/marking.cpp [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+   {
+      ElementTransformation *Tr = pmesh.GetElementTransformation(i);
+      const IntegrationRule &ir =
+         IntRulesLo.Get(pmesh.GetElementBaseGeometry(i), 4*Tr->OrderJ());
+      ls_func.GetValues(i, ir, vals);
+
+      int count = 0;
+      for (int j = 0; j < ir.GetNPoints(); j++)
+      {
+         if (vals(j) <= 0.) { count++; }
```
Comment (dylan-copeland) [inline miniapps/shifted/marking.cpp:36]: I wonder if a strict condition like this will work in general, or if a tolerance should be used (`vals(j) < tol`). If there is round-off or numerical error, points on the boundary of an element may not be well-defined by this condition without a tolerance (if the integration rule has boundary points or points very close to the boundary). I am not familiar with `HeatDistanceSolver`, but could that have numerical error that may require using a tolerance? If a tolerance is necessary, defining it in general for any problem may be challenging.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: miniapps/shifted/diffusion.cpp [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+      }
+      char vishost[] = "localhost";
+      int  visport   = 19916, s = 350;
+      socketstream sol_sock;
+      common::VisualizeField(sol_sock, vishost, visport, elem_marker_gf,
+                             "Element Flags", 0, 0, s, s, "Rjmpc");
+   }
+
+   // Get a list of dofs associated with shifted boundary faces.
+   Array<int> sbm_dofs; // Array of dofs on SBM faces
```
Comment (tzanio) [inline miniapps/shifted/diffusion.cpp:201]: Define SBM

File: miniapps/shifted/diffusion.cpp [hunk 19]
```diff
@@ -0,0 +1,474 @@
...
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+//  ---------------------------------------------------------------------------
+//  Shifted Boundary Diffusion Miniapp: Finite element immersed boundary solver
+//  ---------------------------------------------------------------------------
```
Comment (tzanio) [inline miniapps/shifted/diffusion.cpp:14]: Add to `doc/CodeDocumentation.dox`? (See https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-request-checklist)

**Review by tzanio — APPROVED** [review]
Very cool, and very impressive work @kmittal2 and @vladotomov! 🚀 

@bslazarov and @dylan-copeland -- thank you for the review!

File: fem/lininteg.hpp [hunk 20]
```diff
@@ -35,6 +35,10 @@ class LinearFormIntegrator
    virtual void AssembleRHSElementVect(const FiniteElement &el,
                                        FaceElementTransformations &Tr,
                                        Vector &elvect);
+   virtual void AssembleRHSElementVect(const FiniteElement &el1,
+                                       const FiniteElement &el2,
+                                       FaceElementTransformations &Tr,
+                                       Vector &elvect);
```
Comment (v-dobrev) [inline fem/lininteg.hpp:41]: This addition causes the following warnings from `nvcc`:
```console
.../fem/lininteg.hpp(154): warning: overloaded virtual function "mfem::LinearFormIntegrator::AssembleRHSElementVect" is only partially overridden in class "mfem::BoundaryLFIntegrator"

.../fem/lininteg.hpp(403): warning: overloaded virtual function "mfem::LinearFormIntegrator::AssembleRHSElementVect" is only partially overridden in class "mfem::BoundaryFlowIntegrator"

.../fem/lininteg.hpp(440): warning: overloaded virtual function "mfem::LinearFormIntegrator::AssembleRHSElementVect" is only partially overridden in class "mfem::DGDirichletLFIntegrator"

.../fem/lininteg.hpp(480): warning: overloaded virtual function "mfem::LinearFormIntegrator::AssembleRHSElementVect" is only partially overridden in class "mfem::DGElasticityDirichletLFIntegrator"

```

Comment (tzanio) [reply fem/lininteg.hpp:41]: @kmittal2, can you address this?

It would be nice if we can catch these things automatically.

File: miniapps/shifted/diffusion.cpp [hunk 21]
```diff
@@ -0,0 +1,475 @@
...
+                  "Nitsche penalty parameter (~1 for 2D, ~10 for 3D).");
+   args.AddOption(&include_cut_cell, "-cut", "--cut", "-no-cut-cell",
+                  "--no-cut-cell",
+                  "Include or not include elements cut by true boundary.");
+
+   args.Parse();
+   if (!args.Good())
+   {
+      args.PrintUsage(cout);
+      return 1;
```
Comment (v-dobrev) [inline miniapps/shifted/diffusion.cpp:113]: Print from rank 0 only.

Missing `MPI_Finalize()`.

File: miniapps/shifted/diffusion.cpp [hunk 22]
```diff
@@ -0,0 +1,475 @@
...
+   }
+   args.PrintOptions(cout);
```
Comment (v-dobrev) [inline miniapps/shifted/diffusion.cpp:115]: Print from rank 0 only.

File: miniapps/shifted/diffusion.cpp [hunk 23]
```diff
@@ -0,0 +1,475 @@
...
+
+   // Enable hardware devices such as GPUs, and programming models such as CUDA,
+   // OCCA, RAJA and OpenMP based on command line options.
+   Device device("cpu");
+   device.Print();
```
Comment (v-dobrev) [inline miniapps/shifted/diffusion.cpp:120]: Print from rank 0 only.

**Comment by tzanio** [comment]
@kmittal2 can you address @v-dobrev's comments above?
